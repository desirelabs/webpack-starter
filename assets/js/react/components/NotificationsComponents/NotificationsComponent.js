// @flow
import React from 'react'
import axios from 'axios'
import { Dropdown, ButtonToolbar } from 'react-bootstrap'
import Modal from '../components/ModalComponent'

import NotificationsList from './NotificationsList'
import NotificationsHeader from './NotificationsHeader'
import NotificationsFooter from './NotificationsFooter'
import Badge from '../components/BadgeComponent'

import notifIcon from './assets/notifications-icon.svg'

type Props = {}

type notification = {
  createdAt: string,
  dismissed: boolean,
  id: number,
  routeName: string,
  routeVars: Object,
  title: string
}

type State = {
  showModal: boolean,
  modalAction: Function,
  headerMessage: string,
  bodyMessage: string,
  notifications: Array<notification>
}

class NotificationsComponent extends React.Component<Props, State> {
  constructor () {
    super()
    this.state = {
      showModal: false,
      modalAction: () => {},
      headerMessage: '',
      bodyMessage: '',
      notifications: []
    }
  }

  dismissAll = () => {
    axios.post(`http://localhost:3000/notifications/`, { dismiss: true }).then(() => {
      this.setState({
        notifications: this.state.notifications.filter(notif => {
          notif.dismissed = true
          return notif
        })
      })
    }).catch(error => {
      console.log('Error', error.message)
    })
  }

  deleteAll = () => {
    axios.post(`http://localhost:3000/notifications/`).then(response => {
      this.setState({
        notifications: []
      }, this.resetModal())
    }).catch(error => {
      console.log('Error', error.message)
      this.resetModal()
    })
  }

  dismissNotification = (id: number) => {
    this.state.notifications.filter(notif => {
      if (notif.id === id) {
        axios.put(`http://localhost:3000/notifications/${id}`,
          {
            ...notif,
            dismissed: !notif.dismissed
          }
        ).then(() => {
          this.setState({
            notifications: this.state.notifications.filter(notif => {
              if (notif.id === id) {
                notif.dismissed = !notif.dismissed
              }
              return notif
            })
          })
        })
      }
    })
  }

  getNotifications () {
    axios.get('http://localhost:3000/notifications', {
      headers: {'Accept': 'application/json'}
    }).then(response => {
      this.setState({
        notifications: response.data
      })
    }).catch(error => {
      console.log('Error : ', error.message)
    })
  }

  deleteNotification = (id: number) => {
    axios.delete(`http://localhost:3000/notifications/${id}`).then(response => {
      this.setState({
        notifications: this.state.notifications.filter(notif => notif.id !== id)
      }, this.resetModal())
    }).catch(error => {
      console.log('Error', error.message)
      this.resetModal()
    })
  }

  onNotifClick = (id: number) => {
    window.location = `http://localhost:3000/notifications/${id}`
  }

  resetModal = () => {
    this.setState({
      showModal: false,
      modalAction: () => {},
      headerMessage: '',
      bodyMessage: ''
    })
  }

  configModal = (headerMessage: string, bodyMessage: string, action: Function) => {
    this.setState({
      showModal: true,
      headerMessage: headerMessage,
      bodyMessage: bodyMessage,
      modalAction: action
    })
  }

  componentWillMount () {
    this.getNotifications()
  }

  render () {
    return (
      <ButtonToolbar style={notificationsWrapper}>
        <Dropdown id='dropdown-custom-1'>
          <Dropdown.Toggle noCaret style={dropdownButton}>
            <img src={notifIcon} alt='Notifications' width='40' height='40' />
            {this.state.notifications.filter(notif => !notif.dismissed).length > 0 &&
            <Badge count={this.state.notifications.filter(notif => !notif.dismissed).length} style={{}} />
            }
          </Dropdown.Toggle>
          <Dropdown.Menu className='dropdown-menu animated bounceInDown' style={{borderRadius: '5px', overflow: 'hidden'}}>
            <div>
              <NotificationsHeader count={this.state.notifications.filter(notif => !notif.dismissed).length} />
              <NotificationsList notifications={this.state.notifications} onNotifClick={this.onNotifClick} onDismiss={this.dismissNotification} onDelete={(id) => this.configModal('Suppression', 'Suppprimer ce message ?', () => this.deleteNotification(id))} />
              <NotificationsFooter onDismissAll={this.dismissAll} onDeleteAll={() => this.configModal('Suppression de toutes les notifications', 'Supprimer tous les messages ?', () => this.deleteAll())} />
            </div>
          </Dropdown.Menu>
        </Dropdown>
        <Modal showModal={this.state.showModal} headerMessage={this.state.headerMessage} bodyMessage={this.state.bodyMessage} onCancel={this.resetModal} onConfirm={this.state.modalAction} />
      </ButtonToolbar>
    )
  }
}

const notificationsWrapper = {
  minHeight: '90px',
  display: 'flex',
  alignItems: 'center'
}

const dropdownButton = {
  border: 'none',
  backgroundColor: 'transparent',
  boxShadow: 'none',
  marginRight: '-15px'
}

export default NotificationsComponent
