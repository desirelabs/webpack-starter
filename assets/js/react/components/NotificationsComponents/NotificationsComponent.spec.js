import React from 'react'
import { configure, shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-15'

import NotificationsComponent from './NotificationsComponent'

configure({adapter: new Adapter()})

describe('Testing NotificationsComponent', () => {
  /**
   * This should call getNotifications
   */
  test('getNotifications should be called', () => {
    global.Routing = { generate: () => true }
    const getNotificationsMock = () => true
    let wrapper = shallow(<NotificationsComponent />)
    wrapper.getNotifications = getNotificationsMock
    expect(wrapper).toMatchSnapshot()
  })
})
