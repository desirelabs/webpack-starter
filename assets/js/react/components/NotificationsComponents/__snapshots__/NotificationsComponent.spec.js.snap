// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing NotificationsComponent getNotifications should be called 1`] = `
ShallowWrapper {
  "getNotifications": [Function],
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NotificationsComponent />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "bsClass": "btn-toolbar",
      "children": Array [
        <Uncontrolled(Dropdown)
          id="dropdown-custom-1"
        >
          <DropdownToggle
            bsClass="dropdown-toggle"
            bsRole="toggle"
            noCaret={true}
            open={false}
            style={
              Object {
                "backgroundColor": "transparent",
                "border": "none",
                "boxShadow": "none",
                "marginRight": "-15px",
              }
            }
            useAnchor={false}
          >
            <img
              alt="Notifications"
              height="40"
              src="test-file-stub"
              width="40"
            />
          </DropdownToggle>
          <DropdownMenu
            bsClass="dropdown-menu"
            bsRole="menu"
            className="dropdown-menu animated bounceInDown"
            pullRight={false}
            style={
              Object {
                "borderRadius": "5px",
                "overflow": "hidden",
              }
            }
          >
            <div>
              <NotificationsHeader
                count={0}
              />
              <NotificationsList
                notifications={Array []}
                onDelete={[Function]}
                onDismiss={[Function]}
                onNotifClick={[Function]}
              />
              <NotificationsFooter
                onDeleteAll={[Function]}
                onDismissAll={[Function]}
              />
            </div>
          </DropdownMenu>
        </Uncontrolled(Dropdown)>,
        <ModalComponent
          bodyMessage=""
          headerMessage=""
          onCancel={[Function]}
          onConfirm={[Function]}
          showModal={false}
        />,
      ],
      "style": Object {
        "alignItems": "center",
        "display": "flex",
        "minHeight": "90px",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <DropdownToggle
              bsClass="dropdown-toggle"
              bsRole="toggle"
              noCaret={true}
              open={false}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "border": "none",
                  "boxShadow": "none",
                  "marginRight": "-15px",
                }
              }
              useAnchor={false}
            >
              <img
                alt="Notifications"
                height="40"
                src="test-file-stub"
                width="40"
              />
            </DropdownToggle>,
            <DropdownMenu
              bsClass="dropdown-menu"
              bsRole="menu"
              className="dropdown-menu animated bounceInDown"
              pullRight={false}
              style={
                Object {
                  "borderRadius": "5px",
                  "overflow": "hidden",
                }
              }
            >
              <div>
                <NotificationsHeader
                  count={0}
                />
                <NotificationsList
                  notifications={Array []}
                  onDelete={[Function]}
                  onDismiss={[Function]}
                  onNotifClick={[Function]}
                />
                <NotificationsFooter
                  onDeleteAll={[Function]}
                  onDismissAll={[Function]}
                />
              </div>
            </DropdownMenu>,
          ],
          "id": "dropdown-custom-1",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "bsClass": "dropdown-toggle",
              "bsRole": "toggle",
              "children": Array [
                <img
                  alt="Notifications"
                  height="40"
                  src="test-file-stub"
                  width="40"
                />,
                false,
              ],
              "noCaret": true,
              "open": false,
              "style": Object {
                "backgroundColor": "transparent",
                "border": "none",
                "boxShadow": "none",
                "marginRight": "-15px",
              },
              "useAnchor": false,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "alt": "Notifications",
                  "height": "40",
                  "src": "test-file-stub",
                  "width": "40",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              false,
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "bsClass": "dropdown-menu",
              "bsRole": "menu",
              "children": <div>
                <NotificationsHeader
                  count={0}
                />
                <NotificationsList
                  notifications={Array []}
                  onDelete={[Function]}
                  onDismiss={[Function]}
                  onNotifClick={[Function]}
                />
                <NotificationsFooter
                  onDeleteAll={[Function]}
                  onDismissAll={[Function]}
                />
              </div>,
              "className": "dropdown-menu animated bounceInDown",
              "pullRight": false,
              "style": Object {
                "borderRadius": "5px",
                "overflow": "hidden",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <NotificationsHeader
                    count={0}
                  />,
                  <NotificationsList
                    notifications={Array []}
                    onDelete={[Function]}
                    onDismiss={[Function]}
                    onNotifClick={[Function]}
                  />,
                  <NotificationsFooter
                    onDeleteAll={[Function]}
                    onDismissAll={[Function]}
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "count": 0,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "notifications": Array [],
                    "onDelete": [Function],
                    "onDismiss": [Function],
                    "onNotifClick": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "onDeleteAll": [Function],
                    "onDismissAll": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "bodyMessage": "",
          "headerMessage": "",
          "onCancel": [Function],
          "onConfirm": [Function],
          "showModal": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "bsClass": "btn-toolbar",
        "children": Array [
          <Uncontrolled(Dropdown)
            id="dropdown-custom-1"
          >
            <DropdownToggle
              bsClass="dropdown-toggle"
              bsRole="toggle"
              noCaret={true}
              open={false}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "border": "none",
                  "boxShadow": "none",
                  "marginRight": "-15px",
                }
              }
              useAnchor={false}
            >
              <img
                alt="Notifications"
                height="40"
                src="test-file-stub"
                width="40"
              />
            </DropdownToggle>
            <DropdownMenu
              bsClass="dropdown-menu"
              bsRole="menu"
              className="dropdown-menu animated bounceInDown"
              pullRight={false}
              style={
                Object {
                  "borderRadius": "5px",
                  "overflow": "hidden",
                }
              }
            >
              <div>
                <NotificationsHeader
                  count={0}
                />
                <NotificationsList
                  notifications={Array []}
                  onDelete={[Function]}
                  onDismiss={[Function]}
                  onNotifClick={[Function]}
                />
                <NotificationsFooter
                  onDeleteAll={[Function]}
                  onDismissAll={[Function]}
                />
              </div>
            </DropdownMenu>
          </Uncontrolled(Dropdown)>,
          <ModalComponent
            bodyMessage=""
            headerMessage=""
            onCancel={[Function]}
            onConfirm={[Function]}
            showModal={false}
          />,
        ],
        "style": Object {
          "alignItems": "center",
          "display": "flex",
          "minHeight": "90px",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <DropdownToggle
                bsClass="dropdown-toggle"
                bsRole="toggle"
                noCaret={true}
                open={false}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "border": "none",
                    "boxShadow": "none",
                    "marginRight": "-15px",
                  }
                }
                useAnchor={false}
              >
                <img
                  alt="Notifications"
                  height="40"
                  src="test-file-stub"
                  width="40"
                />
              </DropdownToggle>,
              <DropdownMenu
                bsClass="dropdown-menu"
                bsRole="menu"
                className="dropdown-menu animated bounceInDown"
                pullRight={false}
                style={
                  Object {
                    "borderRadius": "5px",
                    "overflow": "hidden",
                  }
                }
              >
                <div>
                  <NotificationsHeader
                    count={0}
                  />
                  <NotificationsList
                    notifications={Array []}
                    onDelete={[Function]}
                    onDismiss={[Function]}
                    onNotifClick={[Function]}
                  />
                  <NotificationsFooter
                    onDeleteAll={[Function]}
                    onDismissAll={[Function]}
                  />
                </div>
              </DropdownMenu>,
            ],
            "id": "dropdown-custom-1",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "bsClass": "dropdown-toggle",
                "bsRole": "toggle",
                "children": Array [
                  <img
                    alt="Notifications"
                    height="40"
                    src="test-file-stub"
                    width="40"
                  />,
                  false,
                ],
                "noCaret": true,
                "open": false,
                "style": Object {
                  "backgroundColor": "transparent",
                  "border": "none",
                  "boxShadow": "none",
                  "marginRight": "-15px",
                },
                "useAnchor": false,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "Notifications",
                    "height": "40",
                    "src": "test-file-stub",
                    "width": "40",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                false,
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "bsClass": "dropdown-menu",
                "bsRole": "menu",
                "children": <div>
                  <NotificationsHeader
                    count={0}
                  />
                  <NotificationsList
                    notifications={Array []}
                    onDelete={[Function]}
                    onDismiss={[Function]}
                    onNotifClick={[Function]}
                  />
                  <NotificationsFooter
                    onDeleteAll={[Function]}
                    onDismissAll={[Function]}
                  />
                </div>,
                "className": "dropdown-menu animated bounceInDown",
                "pullRight": false,
                "style": Object {
                  "borderRadius": "5px",
                  "overflow": "hidden",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <NotificationsHeader
                      count={0}
                    />,
                    <NotificationsList
                      notifications={Array []}
                      onDelete={[Function]}
                      onDismiss={[Function]}
                      onNotifClick={[Function]}
                    />,
                    <NotificationsFooter
                      onDeleteAll={[Function]}
                      onDismissAll={[Function]}
                    />,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "count": 0,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "notifications": Array [],
                      "onDelete": [Function],
                      "onDismiss": [Function],
                      "onNotifClick": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": null,
                    "nodeType": "function",
                    "props": Object {
                      "onDeleteAll": [Function],
                      "onDismissAll": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "div",
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "bodyMessage": "",
            "headerMessage": "",
            "onCancel": [Function],
            "onConfirm": [Function],
            "showModal": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
